x-pack/plugins/fleet/server/services/epm/elasticsearch/datastream_ilm/install.ts

```diff
@@ -0,0 +1,137 @@
+  if (installation) {
+    previousInstalledIlmEsAssets = installation.installed_es.filter(
+      ({ type, id }) => type === ElasticsearchAssetType.dataStreamIlmPolicy
+    );
+  }
+
+  // delete all previous ilm
+  await deleteIlms(
+    callCluster,
+    previousInstalledIlmEsAssets.map((asset) => asset.id)
+  );
+  // install the latest dataset
+  const dataStreams = registryPackage.data_streams;
+  if (!dataStreams?.length) return [];
+  const dataStreamIlmPaths = paths.filter((path) => isDataStreamIlm(path));
+  let installedIlms: EsAssetReference[] = [];
+  if (dataStreamIlmPaths.length > 0) {
+    const ilmPathDatasets = dataStreams.reduce<IlmPathDataset[]>((acc, dataStream) => {
+      dataStreamIlmPaths.forEach((path) => {
+        if (isDatasetIlm(path, dataStream.path)) {
+          acc.push({ path, dataStream });
+        }
+      });
+      return acc;
+    }, []);
+
+    const ilmRefs = ilmPathDatasets.reduce<EsAssetReference[]>((acc, ilmPathDataset) => {
+      if (ilmPathDataset) {
+        acc.push({
+          id: getIlmNameForInstallation(ilmPathDataset),
+          type: ElasticsearchAssetType.dataStreamIlmPolicy,
+        });
+      }
+      return acc;
+    }, []);
+
+    await saveInstalledEsRefs(savedObjectsClient, registryPackage.name, ilmRefs);
+
+    const ilmInstallations: IlmInstallation[] = ilmPathDatasets.map(
+      (ilmPathDataset: IlmPathDataset) => {
+        return {
+          installationName: getIlmNameForInstallation(ilmPathDataset),
+          content: getAsset(ilmPathDataset.path).toString('utf-8'),
+        };
+      }
+    );
+
+    const installationPromises = ilmInstallations.map(async (ilmInstallation) => {
+      return handleIlmInstall({ callCluster, ilmInstallation });
+    });
+
+    installedIlms = await Promise.all(installationPromises).then((results) => results.flat());
+  }
+
+  if (previousInstalledIlmEsAssets.length > 0) {
+    const currentInstallation = await getInstallation({
+      savedObjectsClient,
+      pkgName: registryPackage.name,
+    });
+
+    // remove the saved object reference
+    await deleteIlmRefs(
+      savedObjectsClient,
+      currentInstallation?.installed_es || [],
+      registryPackage.name,
+      previousInstalledIlmEsAssets.map((asset) => asset.id),
+      installedIlms.map((installed) => installed.id)
+    );
+  }
+  return installedIlms;
+};
+
+async function handleIlmInstall({
+  callCluster,
+  ilmInstallation,
+}: {
+  callCluster: CallESAsCurrentUser;
+  ilmInstallation: IlmInstallation;
+}): Promise<EsAssetReference> {
+  await callCluster('transport.request', {
+    method: 'PUT',
+    path: `/_ilm/policy/${ilmInstallation.installationName}`,
+    body: ilmInstallation.content,
+  });
+
+  return { id: ilmInstallation.installationName, type: ElasticsearchAssetType.dataStreamIlmPolicy };
+}
+
+const isDataStreamIlm = (path: string) => {
+  return new RegExp('(?<package>.*)/data_stream/(?<dataset>.*)/elasticsearch/ilm/*.*').test(path);
```

Comment:
nit: Do we need the `?<package>` or the `?<dataset>` ? I believe those are so we can extract out those specific fields.

Do we need to escape all the `/`?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions

> Similarly, if you're writing a regular expression literal and need to match a slash ("/"), you need to escape that (otherwise, it terminates the pattern). For instance, to search for the string "/example/" followed by one or more alphabetic characters, you'd use /\/example\/[a-z]+/i—the backslashes before each slash make them literal.

I was playing around with it here: 
https://regex101.com/r/GLnjJN/1
https://regex101.com/r/GLnjJN/2

I think we might need one of these:

```
(?<package>.*)\/data_stream\/(?<dataset>.*)\/elasticsearch\/ilm\/.*
```

```
.*\/data_stream\/.*\/elasticsearch\/ilm\/.*
```

Reply:
Do we *need*? No, probably not. `(?<package>.*)` vs `.*` is adding a named capture group.

It doesn't look like we access or use the groups, so that's unnecessary. Adding the label, though, somewhat acts as in-regex-documentation of what we're looking at. I do like that.

We do not need to escape the slashes here because it is a `new Regexp` constructor using a string. Not using slash-literals.